src/java/org/apache/cassandra/repair/state/WeightedHierarchy.java

```diff
@@ -0,0 +1,63 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.repair.state;
+
+import java.util.concurrent.atomic.AtomicLong;
+
+import org.apache.cassandra.cache.IMeasurableMemory;
+
+/**
+ * A pattern for tracking the weight of a hierarchy of references. Registered objects are treated as though they are
+ * immutable, and are only measured on registration.
+ *
+ * Internal nodes must maintain a reference to the root, in order to notify it of any new registrations that may impact
+ * the total weight of the hierarchy.
+ */
+class WeightedHierarchy
```

Comment:
really not a fan of this class, feels like its more complex than it really needs to be...  why did you go this route vs your original?

Reply:
Since CASSANDRA-18816 introduces more hierarchy (ParticipateState now contains Job, which contains validation and syncs) there are now two of these nested hierarchies, the other one being CoordinatorState-SessionState-JobState. I thought it might be easier to see which objects are roots and which are nested. If it's not your preference I'll remove it and put it back the way it used to be.

Reply:
ill leave to your discretion... not a fan and feel like the old solution of having just the counter in both classes with nested objects incrementing was cleaner and less code...